File: app/src/main/java/com/github/anrimian/musicplayer/ui/playlist_screens/rename/RenamePlayListDialogFragment.java
package com.github.anrimian.musicplayer.ui.playlist_screens.rename;

import static com.github.anrimian.musicplayer.Constants.Arguments.PLAY_LIST_ID_ARG;
import static com.github.anrimian.musicplayer.ui.utils.AndroidUtils.setSoftInputVisible;
import static com.github.anrimian.musicplayer.ui.utils.ViewUtils.setEditableText;

import android.app.AlertDialog;
import android.app.Dialog;
import android.os.Bundle;
import android.text.InputType;
import android.view.LayoutInflater;
import android.view.View;
import android.view.inputmethod.EditorInfo;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ProgressBar;
import android.widget.TextView;

import androidx.annotation.NonNull;

import com.github.anrimian.musicplayer.R;
import com.github.anrimian.musicplayer.databinding.DialogCommonInputBinding;
import com.github.anrimian.musicplayer.di.Components;
import com.github.anrimian.musicplayer.ui.common.error.ErrorCommand;

import moxy.MvpAppCompatDialogFragment;
import moxy.presenter.InjectPresenter;
import moxy.presenter.ProvidePresenter;

public class RenamePlayListDialogFragment extends MvpAppCompatDialogFragment
        implements RenamePlayListView {

    @InjectPresenter
    RenamePlayListPresenter presenter;

    private EditText etPlayListName;
    private TextView tvError;
    private TextView tvProgress;
    private ProgressBar progressBar;

    private Button btnChange;

    public static RenamePlayListDialogFragment newInstance(long playListId) {
        Bundle args = new Bundle();
        args.putLong(PLAY_LIST_ID_ARG, playListId);
        RenamePlayListDialogFragment fragment = new RenamePlayListDialogFragment();
        fragment.setArguments(args);
        return fragment;
    }

    @ProvidePresenter
    RenamePlayListPresenter providePresenter() {
        return Components.getPlayListComponent(getPlayListId()).changePlayListPresenter();
    }

    @NonNull
    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        DialogCommonInputBinding viewBinding = DialogCommonInputBinding.inflate(
                LayoutInflater.from(getContext())
        );
        etPlayListName = viewBinding.etInput;
        tvError = viewBinding.tvError;
        tvProgress = viewBinding.tvProgress;
        progressBar = viewBinding.progressBar;

        View view = viewBinding.getRoot();

        AlertDialog dialog = new AlertDialog.Builder(getActivity())
                .setTitle(R.string.edit_name)
                .setPositiveButton(R.string.change, null)
                .setNegativeButton(R.string.cancel, (dialog1, which) -> {})
                .setView(view)
                .create();
        setSoftInputVisible(dialog.getWindow());
        dialog.show();

        etPlayListName.setImeOptions(EditorInfo.IME_ACTION_DONE);
        etPlayListName.setRawInputType(InputType.TYPE_CLASS_TEXT);
        etPlayListName.setOnEditorActionListener((v, actionId, event) -> {
            onCompleteButtonClicked();
            return true;
        });
        etPlayListName.requestFocus();

        btnChange = dialog.getButton(AlertDialog.BUTTON_POSITIVE);
        btnChange.setOnClickListener(v -> onCompleteButtonClicked());

        return dialog;
    }

    @Override
    public void showProgress() {
        btnChange.setEnabled(false);
        etPlayListName.setEnabled(false);
        tvError.setVisibility(View.GONE);
        tvProgress.setVisibility(View.VISIBLE);
        progressBar.setVisibility(View.VISIBLE);
    }

    @Override
    public void showInputState() {
        btnChange.setEnabled(true);
        etPlayListName.setEnabled(true);
        tvError.setVisibility(View.GONE);
        tvProgress.setVisibility(View.GONE);
        progressBar.setVisibility(View.GONE);
    }

    @Override
    public void showError(ErrorCommand errorCommand) {
        btnChange.setEnabled(true);
        etPlayListName.setEnabled(true);
        tvProgress.setVisibility(View.GONE);
        progressBar.setVisibility(View.GONE);
        tvError.setVisibility(View.VISIBLE);
        tvError.setText(getString(R.string.change_playlist_name_error_template, errorCommand.getMessage()));
    }

    @Override
    public void showPlayListName(String initialName) {
        if (etPlayListName.length() == 0) {
            setEditableText(etPlayListName, initialName);
        }
    }

    @Override
    public void closeScreen() {
        dismissAllowingStateLoss();
    }

    private void onCompleteButtonClicked() {
        presenter.onCompleteInputButtonClicked(etPlayListName.getText().toString());
    }

    private long getPlayListId() {
        return requireArguments().getLong(PLAY_LIST_ID_ARG);
    }
}

File: app/src/main/java/com/github/anrimian/musicplayer/ui/playlist_screens/rename/RenamePlayListDialogFragment.kt
package com.github.anrimian.musicplayer.ui.playlist_screens.rename

import android.app.AlertDialog
import android.app.Dialog
import android.os.Bundle
import android.text.InputType
import android.view.LayoutInflater
import android.view.View
import android.view.inputmethod.EditorInfo
import android.widget.Button
import com.github.anrimian.musicplayer.Constants
import com.github.anrimian.musicplayer.R
import com.github.anrimian.musicplayer.databinding.DialogCommonInputBinding
import com.github.anrimian.musicplayer.di.Components
import com.github.anrimian.musicplayer.ui.common.error.ErrorCommand
import com.github.anrimian.musicplayer.ui.utils.AndroidUtils
import com.github.anrimian.musicplayer.ui.utils.ViewUtils
import moxy.MvpAppCompatDialogFragment
import moxy.ktx.moxyPresenter

fun newRenamePlaylistDialog(playListId: Long) = RenamePlayListDialogFragment().apply {
    arguments = Bundle().apply {
        putLong(Constants.Arguments.PLAY_LIST_ID_ARG, playListId)
    }
}

class RenamePlayListDialogFragment : MvpAppCompatDialogFragment(), RenamePlayListView {

    private val presenter by moxyPresenter {
        Components.getPlayListComponent(getPlayListId()).changePlayListPresenter()
    }

    private lateinit var binding: DialogCommonInputBinding

    private lateinit var btnChange: Button

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        binding = DialogCommonInputBinding.inflate(LayoutInflater.from(context))
        
        val dialog = AlertDialog.Builder(activity)
            .setTitle(R.string.edit_name)
            .setPositiveButton(R.string.change, null)
            .setNegativeButton(R.string.cancel) { _, _ -> }
            .setView(binding.root)
            .create()
        AndroidUtils.setSoftInputVisible(dialog.window)
        dialog.show()
        
        binding.etInput.imeOptions = EditorInfo.IME_ACTION_DONE
        binding.etInput.setRawInputType(InputType.TYPE_CLASS_TEXT)
        binding.etInput.setOnEditorActionListener { _, _, _ -> 
            onCompleteButtonClicked()
            true
        }
        binding.etInput.requestFocus()
        btnChange = dialog.getButton(AlertDialog.BUTTON_POSITIVE)
        btnChange.setOnClickListener { onCompleteButtonClicked() }
        return dialog
    }

    override fun showProgress() {
        btnChange.isEnabled = false
        binding.etInput.isEnabled = false
        binding.tvError.visibility = View.GONE
        binding.tvProgress.visibility = View.VISIBLE
        binding.progressBar.visibility = View.VISIBLE
    }

    override fun showInputState() {
        btnChange.isEnabled = true
        binding.etInput.isEnabled = true
        binding.tvError.visibility = View.GONE
        binding.tvProgress.visibility = View.GONE
        binding.progressBar.visibility = View.GONE
    }

    override fun showError(errorCommand: ErrorCommand) {
        btnChange.isEnabled = true
        binding.etInput.isEnabled = true
        binding.tvProgress.visibility = View.GONE
        binding.progressBar.visibility = View.GONE
        binding.tvError.visibility = View.VISIBLE
        binding.tvError.text = getString(
            R.string.change_playlist_name_error_template,
            errorCommand.message
        )
    }

    override fun showPlayListName(initialName: String?) {
        if (binding.etInput.length() == 0) {
            ViewUtils.setEditableText(binding.etInput, initialName)
        }
    }

    override fun closeScreen() {
        dismissAllowingStateLoss()
    }

    private fun onCompleteButtonClicked() {
        presenter.onCompleteInputButtonClicked(binding.etInput.text.toString())
    }

    private fun getPlayListId() = requireArguments().getLong(Constants.Arguments.PLAY_LIST_ID_ARG)

}